server: TestSpanStatsFanOutFaultTolerance failed

server.TestSpanStatsFanOutFaultTolerance [failed](https://tanzanite.cluster.engflow.com/invocations/default/74e9f517-8dbe-425f-a72f-33d8bfac47fb?testReportRun=2&testReportShard=15&testReportAttempt=1#targets-Ly9wa2cvc2VydmVyOnNlcnZlcl90ZXN0) on release-24.2 @ [b8b31aaa48e6ab9e4d4c301e1c5a739901cb72b8](https://github.com/cockroachdb/cockroach/commits/b8b31aaa48e6ab9e4d4c301e1c5a739901cb72b8):


```
=== RUN   TestSpanStatsFanOutFaultTolerance
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestSpanStatsFanOutFaultTolerance558590717
    test_log_scope.go:81: use -show-logs to present logs inline
    span_stats_test.go:296: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/server_test/pkg/server/span_stats_test.go:296
        	            				github.com/cockroachdb/cockroach/pkg/server_test/pkg/server/span_stats_test.go:331
        	Error:      	Not equal: 
        	            	expected: 3
        	            	actual  : 4
        	Test:       	TestSpanStatsFanOutFaultTolerance
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestSpanStatsFanOutFaultTolerance558590717
--- FAIL: TestSpanStatsFanOutFaultTolerance (267.52s)
=== RUN   TestSpanStatsFanOutFaultTolerance/partial-fanout-failure
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    testing.go:1576: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
    --- FAIL: TestSpanStatsFanOutFaultTolerance/partial-fanout-failure (127.13s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=15</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-prs
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSpanStatsFanOutFaultTolerance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-41180

--- a/pkg/server/span_stats_test.go
+++ b/pkg/server/span_stats_test.go
@@ -402,7 +402,13 @@ func TestSpanStatsMultiStoreReplicationOff(t *testing.T) {
 func TestSpanStatsFanOutFaultTolerance(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.UnderStressWithIssue(t, 108534)
+
+	// This test simulates a specific set of failures and verifies that
+	// the response collects them appropriately. It is flaky under stress
+	// and race.
+	skip.UnderStress(t)
+	skip.UnderRace(t)
+
 	ctx := context.Background()
 	const numNodes = 5
 
@@ -410,7 +416,7 @@ func TestSpanStatsFanOutFaultTolerance(t *testing.T) {
 		name         string
 		dialCallback func(nodeID roachpb.NodeID) error
 		nodeCallback func(ctx context.Context, nodeID roachpb.NodeID) error
-		assertions   func(res *roachpb.SpanStatsResponse)
+		assertions   func(t *testing.T, res *roachpb.SpanStatsResponse)
 	}
 
 	containsError := func(errors []string, testString string) bool {
@@ -450,7 +456,7 @@ func TestSpanStatsFanOutFaultTolerance(t *testing.T) {
 				}
 				return nil
 			},
-			assertions: func(res *roachpb.SpanStatsResponse) {
+			assertions: func(t *testing.T, res *roachpb.SpanStatsResponse) {
 				// Expect to still be able to access SpanToStats for keys.EverythingSpan
 				// without panicking, even though there was a failure on every node.
 				require.Equal(t, int64(0), res.SpanToStats[keys.EverythingSpan.String()].TotalStats.LiveCount)
@@ -485,10 +491,13 @@ func TestSpanStatsFanOutFaultTolerance(t *testing.T) {
 				}
 				return nil
 			},
-			assertions: func(res *roachpb.SpanStatsResponse) {
-				require.Greater(t, res.SpanToStats[keys.EverythingSpan.String()].TotalStats.LiveCount, int64(0))
+			assertions: func(t *testing.T, res *roachpb.SpanStatsResponse) {
+				require.Greater(t, res.SpanToStats[keys.EverythingSpan.String()].TotalStats.LiveCount, int64(0),
+					"response contains no stats: %v", res)
 				// 3 nodes could not service their requests.
-				require.Equal(t, 3, len(res.Errors))
+				require.Equal(t, 3, len(res.Errors),
+					"response contains incorrect number of errors: %v", res,
+				)
 
 				require.Equal(t, true, containsError(res.Errors, "error dialing node 1"))
 				require.Equal(t, true, containsError(res.Errors, "kv error on node 3"))
@@ -523,7 +532,7 @@ func TestSpanStatsFanOutFaultTolerance(t *testing.T) {
 			})
 
 			require.NoError(t, err)
-			tCase.assertions(res)
+			tCase.assertions(t, res)
 		})
 	}
 }

